x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsIntegTests.java

```diff
@@ -864,11 +921,41 @@ public void testSnapshotOfSearchableSnapshotIncludesNoDataButCanBeRestored() thr
         }
         assertAcked(client().admin().indices().prepareDelete(restoredIndexName));
 
+        // The repository that contains the cluster snapshot -- may be different from backupRepositoryName if we're only using one repo and
+        // we rename it.
+        final String restoreRepositoryName;
+        if (hasRepositoryUuid && randomBoolean()) {
+            // Re-mount the repository containing the actual data under a different name
+            final RepositoryMetadata repositoryMetadata = client().admin()
+                .cluster()
+                .prepareGetRepositories(repositoryName)
+                .get()
+                .repositories()
+                .get(0);
+
+            // Rename the repository containing the actual data.
+            final String newRepositoryName = randomAlphaOfLength(10).toLowerCase(Locale.ROOT);
+            assertAcked(client().admin().cluster().prepareDeleteRepository(repositoryName));
+            final Settings.Builder settings = Settings.builder().put(repositoryMetadata.settings());
+            if (randomBoolean()) {
+                settings.put("readonly", "true");
+            }
+            assertAcked(
+                clusterAdmin().preparePutRepository(newRepositoryName)
+                    .setVerify(true) // TODO if we're missing repo UUIDs then load all the repository metadata
```

Comment:
One open question: we don't necessarily have the repository UUID loaded if the repository was last created with `?verify=false` and nobody has written to it or tried to list its contents since. In that case you can still restore a snapshot containing searchable snapshot indices but their shards will fail to start with a `RepositoryMissingException` that reports the UUID and original name.

I don't see any particularly nice ways to force the loading of the repository UUID to avoid that from happening, the restore process is all in `server` so we only have sledgehammers like "read all the `RepositoryData from every repository" at our disposal.

Thoughts welcomed on whether this is something to address either here or later, and if so how.

Reply:
It seems like an easy trap to fall into: I think we should try to address this in order to at least avoid searchable snapshot shards to fail at startup. Would it be possible to force loading the UUID before loading the snapshot during pre-recovery phase? It seems that we need to do it anyway if the UUID is not persisted in repositories metadata and we are restarting a node?

Reply:
The repository UUID lives in the cluster metadata so once we know it we'll still have it after a restart (whether rolling or full-cluster).

We can't load it on the data node directly, we would have to tell the master to get the repository metadata. And the trouble is that we don't know which repository to do that for (we wouldn't need the UUID if we knew that) so we basically have to tell the master to get the `RepositoryData` for every UUID-free repository.

It's perhaps not terrible thanks to all the caching; my concern is that caching doesn't always apply. Perhaps it's not too bad. I think I'll address this in a followup tho.

Reply:
@DaveCTurner I think for now you could just force loading the repo data on master by running. `client().admin().cluster().prepareGetSnapshots("repo-name").setVerbose(false). Then once you have the repo that contains your snapshot UUID, the repo uuid should also match. I guess in practice the issue is small here even without caching. It's just loading a single file per repo and even if it's a little larger it should be quick (the CS update per repo is probably more impactful). But mathematically you'll only have to check half the repos on average anyway and that is assuming none of them were ever used or verified so IMO it's not a big deal.

Reply:
Do you mean in this test or in the restore code? The test is fine, it's the restore code that needs fixing, and yeah I'll probably do exactly that call. The question remains, however, what is `"repo-name"`? It's not the name of the repo we're restoring: the important case is restoring a snapshot that contains N searchable snapshot indices, each of which is potentially in a different repo, and we've no idea what any of those repos are called now, so we basically have to get everything.

Reply:
... also how do we tell the restore code to refresh the repository metadata, given that it doesn't technically know anything about searchable snapshots ...